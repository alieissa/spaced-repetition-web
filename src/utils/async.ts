/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
