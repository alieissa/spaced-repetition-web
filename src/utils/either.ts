/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
