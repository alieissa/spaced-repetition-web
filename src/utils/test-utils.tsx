/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
