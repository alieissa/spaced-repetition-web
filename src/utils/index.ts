/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
