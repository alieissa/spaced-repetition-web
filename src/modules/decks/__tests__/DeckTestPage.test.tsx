/** @format */

import { faker } from '@faker-js/faker'
import '@testing-library/jest-dom'
import { act, screen, waitForElementToBeRemoved } from '@testing-library/react'
import { rest } from 'msw'
import { setupServer } from 'msw/node'

import userEvent from '@testing-library/user-event'
import _ from 'lodash'
import { flushPromises, renderWithProviders } from 'src/utils/test-utils'
import { DeckTestPage } from '..'
const mockNavigate = jest.fn(() => ({}))
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

const deckId = faker.string.uuid()
const decksUrl = `${process.env.REACT_APP_API_ENDPOINT}/decks/${deckId}`

// Mounting component this way ensures that the route parameter :deckId
// has id of the deck, i.e. deckId defined above
const mountComponent = () =>
  renderWithProviders(<DeckTestPage />, {
    initialEntries: [`/decks/${deckId}/test`],
    path: 'decks/:deckId/test',
  })

const createCardHandler = (numCards: number = 1) => {
  const createCard = (index: number) => ({
    id: index,
    question: faker.lorem.sentence(5),
    answers: [
      {
        id: faker.string.uuid(),
        content: faker.lorem.sentence(5),
      },
    ],
  })

  return rest.get(decksUrl, (__, res, ctx) => {
    return res(
      ctx.json({
        id: deckId,
        name: faker.word.noun(),
        cards: _.times(numCards, createCard),
      }),
      ctx.status(200),
    )
  })
}
const handlers = [createCardHandler(1)]

const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('DeckTestPage', () => {
  const user = userEvent.setup()

  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = mountComponent()
      expect(asFragment()).toMatchSnapshot()
    })

    it('should display loading', async () => {
      mountComponent()
      const testId = 'deck-test-loading'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
      // See https://davidwcai.medium.com/react-testing-library-and-the-not-wrapped-in-act-errors-491a5629193b
      await waitForElementToBeRemoved(() => screen.queryByTestId(testId))
    })

    it('should display success', async () => {
      mountComponent()

      await act(flushPromises)

      const testId = 'deck-test-success'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
    })

    it('should display failure', async () => {
      server.use(
        rest.get(decksUrl, (__, res, ctx) => {
          return res(ctx.status(500))
        }),
      )
      mountComponent()

      await act(flushPromises)

      const testId = 'deck-test-failure'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
    })
  })

  describe('interaction', () => {
    it('should not display navigation arrows for deck with one card', async () => {
      // Assemble
      server.use(createCardHandler(1))
      mountComponent()
      await act(flushPromises)

      // Assert
      const nextCardBtn = screen.queryByTestId('next-card-btn')
      const previousCardBtn = screen.queryByTestId('previous-card-btn')
      expect(nextCardBtn).not.toBeInTheDocument()
      expect(previousCardBtn).not.toBeInTheDocument()
    })

    it('should disable next on last card', async () => {
      // Assemble
      server.use(createCardHandler(4))
      mountComponent()
      await act(flushPromises)

      // Act
      const nextCardBtn = screen.getByTestId('next-card-btn')
      await act(() => user.click(nextCardBtn))
      await act(() => user.click(nextCardBtn))
      await act(() => user.click(nextCardBtn))

      // Assert
      expect(await screen.findByTestId('test-card-3')).toBeInTheDocument()
      expect(nextCardBtn).toHaveAttribute('disabled')
    })

    it('should disable previous on first card', async () => {
      // Assemble
      server.use(createCardHandler(4))
      mountComponent()
      await act(flushPromises)

      // Act
      const previousCardBtn = screen.getByTestId('previous-card-btn')
      const nextCardBtn = screen.getByTestId('next-card-btn')
      await act(() => user.click(nextCardBtn))
      await act(() => user.click(previousCardBtn))

      // Assert
      expect(await screen.findByTestId('test-card-0')).toBeInTheDocument()
      expect(previousCardBtn).toHaveAttribute('disabled')
    })
  })
})
