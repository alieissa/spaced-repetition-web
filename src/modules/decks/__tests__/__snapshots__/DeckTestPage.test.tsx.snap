// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckTestPage view should render correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="deck-test-success"
  >
    <div
      class="justify-space-between"
      style="padding: 1rem 1rem 1rem 2rem; line-height: 2rem;"
      title="Test Deck"
    >
      <div
        class="align-center"
      >
        <div>
          <button
            class="ui huge icon button"
            style="padding-left: 0px; padding-right: 0px; background: none;"
          >
            <i
              aria-hidden="true"
              class="chevron left icon"
            />
          </button>
        </div>
        <div
          class="ui header"
          style="font-size: 1.5rem; margin-top: 0px; margin-bottom: 0px;"
        >
          Test Deck
        </div>
      </div>
    </div>
    <div
      class="ui fluid card"
      data-testid="test-card-0"
      style="box-shadow: none; border: 1px solid; border-radius: unset; margin-top: 0px;"
    >
      <div
        class="content"
      >
        <div
          class="ui basic segment"
        >
          <h3
            class="ui header"
            style="margin-top: 0px; margin-bottom: 0px;"
          >
            Test Question 0
          </h3>
        </div>
        <form
          class="ui form"
        >
          <div
            class="field"
          >
            <div
              class="ui input"
              color="green"
              style="opacity: 1;"
            >
              <input
                id="answer"
                placeholder="Enter answer here"
                style="border-radius: unset;"
                title="answer"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="flex-row-reverse"
          >
            <button
              class="ui button"
              style="border-radius: 0;"
              type="submit"
            >
              Check
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
